### DISCUSSION TITLE
Resize / split screen on Android

### DISCUSSION THREAD

**Discussion by ptrm** [koreader/koreader#7417] [Ideas]
I would like to benefit from the split screen feature of Android, I also often use popup window apps. Was it ever considered for KOReader to support it? Right now I get a cropped view of unresized content. Dynamic dpi change is what comes to my mind for allowing this.

(Pinging @pazos for the answer)

**Comment by Frenzie** [comment]
I think we pretty much have all the bits and pieces in place besides responding to the requisite (new) Android stuff.

**Comment by ptrm** [comment]
On a 7.8" Onyx this would be a great feature. I take notes as often as read, and now ften resort to my Clara hd to do it side by side ;D

**Comment by pazos** [comment]
See https://github.com/koreader/android-luajit-launcher/issues/238

It is a bit of pain to adjust the viewport based on decoration changes (fullscreen on "normal" mode vs normal in split mode), but the rest should be doable without to much work. In fact the resizing part is actually done for the hw rotation so it should be a matter of removing [this line](https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/AndroidManifest.xml#L39)

**Reply by ptrm** [reply]
Sound great. Though i⁸f `android:resizeableActivity="false"` can be achieved through forcing it in developer settings:
![SmartSelect_20210316-193932_Settings](https://user-images.githubusercontent.com/366258/111362649-5df95a00-868f-11eb-9404-461beb2187e0.jpg)
then it seems to not be enough:

![Screenshot_20210316-194332_Firefox](https://user-images.githubusercontent.com/366258/111363104-eaa41800-868f-11eb-9ab9-5fab76e74c9a.png)

![Screenshot_20210316-194543_Files](https://user-images.githubusercontent.com/366258/111363342-3787ee80-8690-11eb-9df4-4993649564b8.png)

That's on stable 2021.02

**Comment by pazos** [comment]
@ptrm: apps using the android view hierarchy can be forced with a setting like that since views are aware of config changes. That's not our case. 

To get something usable we need to fix:

> width and height in multiwindow mode are calculated based on reported data by window.decorView. Current implementations of android.getScreenWidth() and android.getScreenHeight will return bogus values.

and "wire" `APP_CMD_WINDOW_RESIZED` to a blitbuffer resize. It should work fine as is in most android devices but it is required for chrome os, at least.

I got that working without too much problem, but is not good enough without:

> Disable fullscreen layout while in multiwindow mode. Also need to take constraints into account to calculate width and height

And that's is a PITA for a feature I won't use personally. 

If you're interested I think I can manage to add the changes needed (except the viewport) and you can get an "almost" working version enabling that dev setting to force apps to split mode.

The viewport part would need another soul ;)

**Comment by ptrm** [comment]
> If you're interested I think I can manage to add the changes needed (except the viewport) and you can get an "almost" working version enabling that dev setting to force apps to split mode.

I see. I think this would be a lot anyway and a good start :)
